kernel/src/main/java/org/kframework/parser/ParserUtils.java

```diff
@@ -220,11 +218,16 @@ private String loadDefinitionText(File definitionFile) {
 
         KILtoKORE kilToKore = new KILtoKORE(context, false, kore, leftAssoc);
 
-        HashMap<String, Module> koreModules = new HashMap<>();
-        koreModules.putAll(previousModules.stream().collect(Collectors.toMap(Module::name, m -> m)));
+        scala.collection.mutable.Map<String, Module> koreModules = new scala.collection.mutable.HashMap<>();
+        previousModules.stream().map(m -> koreModules.put(m.name(), m)).collect(Collectors.toList()); // populate koreModules with previousModules
         HashSet<org.kframework.kil.Module> kilModulesSet = new HashSet<>(kilModules);
 
-        Set<Module> finalModules = kilModules.stream().map(m -> kilToKore.apply(m, kilModulesSet, koreModules)).flatMap(m -> Stream.concat(Stream.of(m), Stream.of(koreModules.get(m.name() + "$SYNTAX")))).collect(Collectors.toSet());
+        java.util.List<FlatModule> flatModules = kilModulesSet.stream().map(kilToKore::toFlatModule).sorted(Comparator.comparing(FlatModule::name)).collect(Collectors.toList());
+
+        // accumulate in koreModules the new modules. Use an ordered set to stabilize the error message for circular imports
+        flatModules.stream().map(m -> m.toModule(immutable(new LinkedHashSet<>(flatModules)), koreModules, Seq())).collect(Collectors.toSet());
+        Set<Module> finalModules = mutable(koreModules.values().toSet());
+
```

Comment:
I don't feel comfortable reviewing this piece of code, because I don't understand the intricacies of Java vs. Scala datastructures and mutable vs immutable datastructures in each. @dwightguth can you have a look at this one?

Reply:
@radumereuta just looking at this code, it looks like we'll be calling `m.toModule` on each module in `flatModules`, but it seems to me that this may repeat work because perhaps the first module you call it on ends up already converting the second module you call it on. Is this the case?

Reply:
Yes, that's why I added the caching step below.
```
if (koreModules.contains(this.name))
      return koreModules(this.name)
```

Reply:
But will each subsequent invocation of `toModule` have the updated map from the previous invocation present? I can't quite follow the control flow here, but are you saying that each call of `toModule` will see the updated `koreModules` from the previous call?

Reply:
From discussions on Slack, it looks like `koreModules` is mutable and updated in place, so subsequent invocations will see the prior updated `koreModules` map.